_source/_posts/2022-09-21-angular-dependency-injection.md

```diff
@@ -0,0 +1,296 @@
+
+## Angular's Dependency Injection system
+
+When you use the Angular CLI to generate a service, it automatically adds the code to register the service within Angular's DI system. That's what the `@Injectable()` TypeScript decorator does:
+
+```ts
+@Injectable({
+  providedIn: 'root'
+})
+export class MyService {
+}
+```
+
+Without doing anything else, we can use our dependency in the application by injecting it into the consuming code as a constructor parameter:
+
+```ts
+@Component({
+  // standard component metadata here
+})
+export class MyComponent {
+  constructor(private myService: MyService) { }
+}
+```
+
+In Angular v14, you now have a new option to use the `inject()` function.
+
+Another way to register dependencies is manually providing them through the `providers` array. Different Angular building blocks accept `providers` in the metadata. So you can register a provider like this:
+
+```ts
+@NgModule({
+  imports: // stuff here
+  declarations: // stuff here
+  providers: [
+    MyService
+  ]
+})
+export class AppModule {
+}
+```
+
+There's something else to note, though. Angular's DI system allows you to provide a dependency to different places within the application. While this sounds unnecessarily complicated, it allows you to fine-tune which dependency to use in your consuming code. You can configure the `providers` array in other modules and Angular building blocks, such as components and directives. You'll get the provider you configured closest to the consuming code when you do so.
+
+Now that we have a quick overview of how to provide dependencies, let's review an integral piece of Angular's DI system, Injection Tokens.
+
+## Injection Tokens in Angular
+
+Injection tokens allow us to have values and objects as dependencies. This means we can depend on a string, such as "Hello world!", and objects, which include configuration objects and global variables such as Web APIs. But injection tokens are even more remarkable because we can also create dependencies to constructs that don't have a runtime definition, such as interfaces! Let's take a look at an example using an injection token.
+
+Let's say we want to have a dependency on the bread emoji. You can create the bread emoji, register the token to Angular's DI system, and set up a default value like this:
+
+```ts
+const export BREAD_TOKEN = new InjectionToken<string>('bread', {
+  providedIn: 'root',
+  factory: () => '🍞'
+});
+```
+
+When you want to use the `BREAD_TOKEN`, you'd use the `Inject` decorator:
+
+```ts
+@Component({
+  // standard component metadata here
+})
+export class MyBreadComponent {
+  constructor(@Inject(BREAD_TOKEN) private bread: string) { }
+}
+```
+
+Now we can access the 🍞emoji from within the component! This might not seem like a big deal, but having injection tokens as a means to represent values and interfaces as dependencies is enormous! And it sets us up to leverage the power of Angular's DI system.
```

Comment:
```suggestion
Now we can access the 🍞 emoji from within the component! This might not seem like a big deal, but having injection tokens as a means to represent values and interfaces as dependencies is enormous! And it sets us up to leverage the power of Angular's DI system.
```
added a space